api/review_checklist.md

```diff
@@ -0,0 +1,137 @@
+# API Review Checklist
+
+This checklist is intended to be used when reviewing xDS API changes.
+Users who wish to contribute API changes should read this and proactively
+consider the answers to these questions before sending a PR.
+
+## Feature Enablement
+- Are the default values going to cause behavior changes for existing users
+  who do not know about the change and have not updated the resources being
+  served by their control plane?
+  - If yes, do we have some estimate of how many users will be affected?
+  - Why is it justified to change the default behavior, rather than making
+    this feature opt-in?
+    - Some possible justifications include security concerns with existing
+      behavior, or a desire to eliminate legacy behavior.
+  - What is the plan to make this change in a safe way?  For example, is the
+    transition going to be staged over the course of several minor xDS versions?
+  - How will we warn users about this change?
+    - Possible ways to do this include release notes, announcements, warnings
+      from the code, etc.
+- Will users have a way to disable this change if it causes problems?
+  - If not, why do we think that's okay?  (It might be the case that we think
+    it will not actually affect anyone, or no one will care.)
+  - If so, is the mechanism to disable it part of the xDS API, or is it
+    acceptable to have a separate knob for this in the client?  (See also
+    "Genericness" below -- if this is not part of the API, will every xDS
+    client need to add a different knob?  Is consistency across clients
+    important for this?)
+
+## Style
+- If an enum is being introduced, should this be a oneof with empty messages
+  for future API growth?
+- When a new field is introduced for a distinct purpose, should this be a
+  message to allow for future API growth?
+- For time related fields, is a Duration or Timestamp WKT used rather than
+  raw integers for seconds etc?
+- Is the PR aligned with the [API style guidelines](STYLE.md)?
+
+## Validation Rules
+- Is the field mandatory? Does it have the required rule?
+- Is the field numeric? Is it bounded correctly?
+- Is the field repeated? Does it have the correct min/max items numbers? Are
+  the values unique?
+- If a field may eventually be repeated but initially there is a desire to
+  cap it to a single value, consider using repeated with a max length of 1,
+  which is easier to relax in the future.
+
+## Deprecations
+- When a field or enum value is deprecated, according to the minor/patch
+  versioning policy this implies a minor version for support removal. Is the
+  work necessary to add support for the newer replacement field acceptable to
+  known xDS clients in this time frame?
+- No deprecations are allowed when an alternative is "not ready yet" in any
```

Comment:
Haven't we traditionally added a new field and deprecated old ones in the same PR?  I guess better project party is part of pulling the APIs out here.  I think it'd be helpful if before we codify this rule, we have some process around adding a field, making sure the migrations are done in both projects, being able to determine when they're done and having issues filed to do the deprecation.

Reply:
I think we can play this by ear.  Right now, there are only 2 clients, and the API shepherds all talk to each other, so it's fairly easy to coordinate this without much red tape.  Over time, if this becomes more of a problem, we can add more process as appropriate.

Reply:
Can we redact this for now or add an "or get sign off from each project stakeholder for immediate deprecation"?  Given n=2 I think that's less onerous and as we add more stakeholders I think we'd have to move towards the deprecation dance anyway.

Reply:
I've added a sentence indicating to ask the API shepherds if you're not sure about the status of a feature in the major known clients.

Reply:
So I'm on board with playing this by ear, but the documentation this still disallows changes like I mentioned. I want us to be explicit here.  I don't want to pedantic, but  I really want to avoid having codified rules we encourage people in the know to ignore.  If we're Ok playing it by ear, let's remove the restriction that you can't deprecate a field until the replacement is ready in all consuming stacks.  If we're not OK doing this with shephard approval, we should make it clear it is now disallowed to add A and deprecate B in the same PR.

Reply:
Ah, okay, sorry -- I somehow didn't grok that it was the ability to explicitly sign off on exceptions that you wanted to call out.  I've added a clause about that.